### PR TITLE
Populate OD of slave based on ESI file

### PR DESCRIPTION
**Pull request by nicolas71640**
Add an od_generator tool that generate an od_populator.cc source file that populate the CoE::dictionary based on the object dictionnary embedded in the ESI file.

## Limitation
I have handled only basic type of entry

## Tests
To be tested on hardware

### CODE REVIEW

**Comment by nicolas71640** [comment]
@leducp  @Rdk-T

File: examples/master/wdc_foot/wdc_foot_example.cc [hunk 1]
```diff
@@ -143,8 +148,8 @@ int main(int argc, char* argv[])
         std::cerr << e.what() << std::endl;
         return 1;
     }
-
-    for (auto& slave: bus.slaves())
+   
+for (auto& slave: bus.slaves())
```
Comment (leducp) [inline examples/master/wdc_foot/wdc_foot_example.cc:152]: misalignement

File: examples/slave/nuttx/xmc4800/od_populator.cc [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/// This file is auto generated by od_generator.
+
+#include "kickcat/od_populator.h"
+#include "kickcat/CoE/OD.h"
+
+namespace kickcat
+{
+    void populateOD()
```
Comment (leducp) [inline examples/slave/nuttx/xmc4800/od_populator.cc:8]: should be in the CoE sub-namespace

File: lib/slave/include/kickcat/od_populator.h [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+#include <cstdint>
+
+/// Header to be included to populate an Object Dictionnary
+/// The populateOD function can be auto generated via the
+/// od_generator, or can be implemented manually.
+/// For more information please read lib/slave/README.md
+
+namespace kickcat
+{
+    void populateOD();
```
Comment (leducp) [inline lib/slave/include/kickcat/od_populator.h:14]: I feel that OD.h is a better place for this prototype.

File: tools/od_generator.cc [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    return result.str();
+}
+
+std::string dataToString(CoE::DataType dataType, void *data)
+{
+    std::stringstream result;
+    result << "0x" << std::hex;
+    switch (dataType)
+    {
+    case CoE::DataType::INTEGER8:
```
Comment (leducp) [inline tools/od_generator.cc:52]: indentation

File: tools/od_generator.cc [hunk 5]
```diff
@@ -0,0 +1,162 @@
+#include "kickcat/Prints.h"
+#include "kickcat/CoE/EsiParser.h"
+#include <fstream>
+#include <iostream>
+
+using namespace kickcat;
+
+constexpr std::string_view OD_POPULATOR_FILE{"od_populator.cc"};
```
Comment (leducp) [inline tools/od_generator.cc:8]: from here to main(): namespace kickcat

File: tools/od_generator.cc [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+std::string addEnding()
+{
+    std::stringstream result;
+    result << "    }\n}\n";
+    return result.str();
+}
+
+std::string dataToString(CoE::DataType dataType, void *data)
```
Comment (leducp) [inline tools/od_generator.cc:46]: May be useful in OD.h

**Comment by leducp** [comment]
Nice work overall!

File: examples/slave/nuttx/xmc4800/od_populator.cc [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                .index = 0x1018,
+                .code = CoE::ObjectCode::RECORD,
+                .name = "Identity Object"
+            };
+            CoE::addEntry(object,0,8,7,static_cast<CoE::DataType>(5),"Subindex 000",0x4);
+            CoE::addEntry(object,1,32,7,static_cast<CoE::DataType>(7),"Vendor ID",0x29c);
+            CoE::addEntry(object,2,32,7,static_cast<CoE::DataType>(7),"Product code",0x32);
+            CoE::addEntry(object,3,32,7,static_cast<CoE::DataType>(7),"Revision number",0x0);
+            CoE::addEntry(object,4,32,7,static_cast<CoE::DataType>(7),"Serial number",0xcadefeca);
```
Comment (Rdk-T) [inline examples/slave/nuttx/xmc4800/od_populator.cc:21]: Strange, the serial number is different from the foot.xml.
has it just been generated on a different xml ?

Comment (nicolas71640) [reply examples/slave/nuttx/xmc4800/od_populator.cc]: In the ESI, the default data is written in little endian. That's why

**Review by leducp — APPROVED** [review]

**Pull request merged**